src/env_universal_common.rs

```diff
@@ -154,90 +141,92 @@ impl EnvUniversal {
-        // risk here is that if the process crashes or is killed while holding the lock, future
-        // instances of fish will not be able to obtain it. This seems to be a greater risk than that of
-        // data loss on lockless NFS. Users who put their home directory on lockless NFS are playing
-        // with fire anyways.
         // If we have no changes, just load.
         if self.modified.is_empty() {
-            self.load_from_path_narrow(callbacks);
+            let callbacks = self.load_from_path_narrow();
             FLOG!(uvar_file, "universal log no modifications");
-            return false;
+            return (false, callbacks);
         }
 
-        let directory = wdirname(&self.vars_path).to_owned();
-
         FLOG!(uvar_file, "universal log performing full sync");
 
-        // Open the file.
-        let Some(mut vars_file) = self.open_and_acquire_lock() else {
-            FLOG!(uvar_file, "universal log open_and_acquire_lock() failed");
-            return false;
+        let rewrite = |old_file: &File,
+                       tmp_file: &mut File|
+         -> std::io::Result<PotentialUpdate<Option<UniversalReadUpdate>>> {
+            match self.load_from_file(old_file) {
+                Some(potential_update) => {
+                    if potential_update.do_save {
+                        let contents = Self::serialize_with_vars(&potential_update.data.new_vars);
+                        tmp_file.write_all(&contents)?;
+                    }
+                    Ok(PotentialUpdate {
+                        do_save: potential_update.do_save,
+                        data: Some(potential_update.data),
+                    })
+                }
+                None => {
+                    if self.ok_to_save {
+                        let contents = Self::serialize_with_vars(&self.vars);
+                        tmp_file.write_all(&contents)?;
+                    }
+                    Ok(PotentialUpdate {
+                        do_save: self.ok_to_save,
+                        data: None,
+                    })
+                }
+            }
         };
 
-        // Read from it.
-        self.load_from_file(&mut vars_file, callbacks);
-
-        if self.ok_to_save {
-            self.save(&directory)
-        } else {
-            true
+        let real_path = wrealpath(&self.vars_path).unwrap_or_else(|| self.vars_path.clone());
+        match rewrite_via_temporary_file(&real_path, rewrite) {
+            Ok((file_id, potential_update)) => {
+                self.last_read_file_id = file_id;
+                self.ok_to_save = potential_update.do_save;
+                self.modified.clear();
+                match potential_update.data {
+                    Some(UniversalReadUpdate {
+                        export_generation_increment,
+                        new_vars,
+                        callbacks,
+                        ok_to_save: _,
+                    }) => {
```

Comment:
maybe add here `assert_eq!(potential_update.do_save, ok_to_save);`
It's unfortunate that we're duplicating this bool but I guess it's a local optimum